### ISSUE TITLE
libreadline on Debian

### ISSUE THREAD

**Issue by wpoely86** [hpcugent/easybuild-easyconfigs#1315]
For some reason, `libreadline` doesn't link to `libtinfo` (part of `ncurses`). This results in things like:
```
libreadline.so.6: undefined symbol: UP
```
The current fix is to add:
```
preconfigopts = " LDFLAGS='-ltinfo' "
```
to the easyconfig. If I ever find the time, I should look into this and fix it permanently.

**Comment by boegel** [comment]
Hmm, would this hurt on other systems?

**Comment by wpoely86** [comment]
Well, apparently not, because you would have seen it by now. I think it's some weird Debian quirk.
The symbols are in `libncurses` and `libreadline` finds them, else the build would fail. But actually running something with them fails. Are maybe it's an issue with system lua and libreadline.

I should dig deeper, but I don't have the time for the moment.

**Comment by boegel** [comment]
What I meant was, would using the `preconfigopts` you mentioned everywhere hurt the builds that are working now? Where does `-ltinfo` come from?

**Comment by wpoely86** [comment]
It's part of the `ncursus-devel` on most systems. I definitely do not recommend to use it everywhere until we understand exactly what is going on here.

**Comment by boegel** [comment]
OK, keep us posted if you figure out more, and thanks for opening this issue, Google will direct other people suffering from this here, hopefully.

**Comment by sylmarien** [comment]
You're right, Google did its job well!
I currently experience this problem as well on Wheezy (no problem on Squeeze). Since I didn't know how to proceed, I tried to rename the libreadline.so.6 file to something else to see what error I was going to have and truth is... I didn't get any, it worked and proceeded without any problem.

So what is exactly happening ? Is this some optional step or another file is used instead? (another version of libreadline? libreadline from the system if any?)

I'm going to try the fix in the easyconfig as it is (way) cleaner than my current one. Because we only aim to use it on Debian Wheezy for now it should do the trick.

**Comment by wpoely86** [comment]
@sylmarien if you have renamed `libreadline.so.6`, you will end up using the system `libreadline`, and so the problem should be 'fixed' for you.

**Comment by sylmarien** [comment]
@wpoely86 Ok thanks, but isn't it a "bad" solution since it kind of breaks the idea of only working with modules and not being dependent on what is installed on the system?

**Comment by wpoely86** [comment]
yes, it is a far from optimal solution. It should be properly be fixed in the easyconfig. 

On what point did you get this message btw? Are you using Lmod?

**Comment by sylmarien** [comment]
I was getting this message while building ABySS-1.3.6-goolf-1.4.10-Python-2.7.5, I don't remember the exact step, I'm currently reproducing it to be more precise.

Yes, I'm using Lmod.

**Comment by sylmarien** [comment]
I reproduced it. It happens during the sanity check, here is a gist with the logs from the build:
https://gist.github.com/sylmarien/5e262eb307a52c8e1bee

**Comment by wpoely86** [comment]
The issue is this:
With my fix:
```
$ ldd /usr/bin/lua
	linux-vdso.so.1 =>  (0x00007fff9cf77000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00002b7a65230000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00002b7a654b3000)
	libreadline.so.6 => /home/opt/easybuild/software/libreadline/6.3-intel-2015a/lib/libreadline.so.6 (0x00002b7a656b7000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00002b7a6590b000)
	/lib64/ld-linux-x86-64.so.2 (0x00002b7a6500e000)
	libtinfo.so.5 => /lib/x86_64-linux-gnu/libtinfo.so.5 (0x00002b7a65c98000)
	libimf.so => /home/opt/easybuild/software/imkl/11.2.1.133-iimpi-7.2.3-GCC-4.9.2/lib/intel64/libimf.so (0x00002b7a65ec1000)
	libsvml.so => /home/opt/easybuild/software/imkl/11.2.1.133-iimpi-7.2.3-GCC-4.9.2/lib/intel64/libsvml.so (0x00002b7a6637c000)
	libirng.so => /home/opt/easybuild/software/imkl/11.2.1.133-iimpi-7.2.3-GCC-4.9.2/lib/intel64/libirng.so (0x00002b7a6725b000)
	libgcc_s.so.1 => /home/opt/easybuild/software/GCC/4.9.2/lib64/libgcc_s.so.1 (0x00002b7a67462000)
	libintlc.so.5 => /home/opt/easybuild/software/imkl/11.2.1.133-iimpi-7.2.3-GCC-4.9.2/lib/intel64/libintlc.so.5 (0x00002b7a67479000)
```
Without my fix:
```
$ ldd /usr/bin/lua
	linux-vdso.so.1 =>  (0x00007fff6f3ff000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00002af9d299a000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00002af9d2c1d000)
	libreadline.so.6 => /home/opt/easybuild/software/libreadline/6.3-intel-2015a/lib/libreadline.so.6 (0x00002af9d2e21000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00002af9d3075000)
	/lib64/ld-linux-x86-64.so.2 (0x00002af9d2778000)
	libimf.so => /home/opt/easybuild/software/imkl/11.2.1.133-iimpi-7.2.3-GCC-4.9.2/lib/intel64/libimf.so (0x00002af9d3402000)
	libsvml.so => /home/opt/easybuild/software/imkl/11.2.1.133-iimpi-7.2.3-GCC-4.9.2/lib/intel64/libsvml.so (0x00002af9d38bd000)
	libirng.so => /home/opt/easybuild/software/imkl/11.2.1.133-iimpi-7.2.3-GCC-4.9.2/lib/intel64/libirng.so (0x00002af9d479b000)
	libgcc_s.so.1 => /home/opt/easybuild/software/GCC/4.9.2/lib64/libgcc_s.so.1 (0x00002af9d49a3000)
	libintlc.so.5 => /home/opt/easybuild/software/imkl/11.2.1.133-iimpi-7.2.3-GCC-4.9.2/lib/intel64/libintlc.so.5 (0x00002af9d49ba000)
```
Without modules loaded:
```
$ ldd /usr/bin/lua
	linux-vdso.so.1 =>  (0x00007fff66dff000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00002b7ac9ff2000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00002b7aca274000)
	libreadline.so.6 => /lib/x86_64-linux-gnu/libreadline.so.6 (0x00002b7aca479000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00002b7aca6c0000)
	/lib64/ld-linux-x86-64.so.2 (0x00002b7ac9dd0000)
	libtinfo.so.5 => /lib/x86_64-linux-gnu/libtinfo.so.5 (0x00002b7acaa4c000)
```
So the issue is that `lua` links to `libtinfo` and our `libreadline` does not.

**Comment by hajgato** [comment]
Ok, I have a system where 
```
preconfigopts = " LDFLAGS='-ltinfo' "
```
cause a problem:
```
checking whether the C compiler works... no
configure: error: in `/tmp/hajgato/build/libreadline/6.3/GNU-4.9.3-2.25/readline-6.3':
configure: error: C compiler cannot create executables
See `config.log' for more details
```
So its not the solution.

According to my opinion we have to investigate the EB installed ncurses when system ncurses/tlib are present.

**Comment by wpoely86** [comment]
Okay, I digged this one a bit. As it's a complete pain in the butt.

`libreadline` needs some terminal info. It can get this from `libtermcap` or `ncurses`, which are (as far as I understand it) API but not ABI compatible. On most (=all the ones I looked at) systems, `libtermcap` just gives a redirect to `libtinfo`. This libraries is created by splitting a part of the main `ncurses` library because most program only need this and not full `ncurses`. In EB we don't do this and it's one big fat library.  Also, `libreadline` and `ncurses` don't mix. It's next to impossible to use both of them in the same code.

The idea is that, the code that uses `libreadline` can choose which termcap library to use, either `libtermcap` or `ncurses`. But as said before, on most distro's the leads to the same library.

So, my proposed solution is: we explicitly link `libreadline` to `ncurses`. This will definitely solve the problem in this PR but I'm not entirely sure about potential fallout. It seems like most EB stuff already links to `ncurses` when using `libreadline`. I'm still guessing about why this issue does not pop up in RHEL (and derivates) but on Debian we do hit it.

Relevant bugreport: https://bugzilla.redhat.com/show_bug.cgi?id=499837

**Comment by boegel** [comment]
@wpoely86: should we remove the `preconfigopts = " LDFLAGS='-ltinfo' "` the is used in existing `ncurses` easyconfigs in `develop` until we figure this out fully?

It shouldn't have been there since https://github.com/hpcugent/easybuild-easyconfigs/pull/1670 was never merged because of a fear of fallout (which was justified, it seems)?

**Comment by wpoely86** [comment]
Either remove it or add a os dep on `ncurses-devel`. It will be present on all modern systems but @hajgato hit the problem that the `so` is not installed by default.

**Comment by boegel** [comment]
The actual fix we're going to put in place is going to be different than using `preconfigopts = " LDFLAGS='-ltinfo' "`, right? So, my vote goes to trashing it for now.

**Comment by wpoely86** [comment]
well, my proposed solution is:
```python
preconfigopts = " LDFLAGS='-lncurses' " 
```

**Comment by boegel** [comment]
@wpoely86: can you look into a PR that trashes the `-ltinfo` trick, and then another with your proposed solution (on one libreadine easyconfig, for now)? so we can test the latter at lenght

**Comment by hajgato** [comment]
Why dont we provide `libtermcap` as well? As Debian does just a symlink to `libncurses`?

**Comment by hajgato** [comment]
^ as a module or in `ncurses`?

**Comment by boegel** [comment]
@hajgato: you mean, symlink `libncurses.a` to `libtermcap.a` in the `ncurses` module?

**Comment by wpoely86** [comment]
no, we don't need to prove `libtermcap` is will be present on the system. We could create a `libtinfo` symlink to `libncurses` in the ncurses module. But that might really interfere with systems tools...

**Comment by hajgato** [comment]
maybe.  Anyway, readline has a ncures dep, so `preconfigopts = " LDFLAGS='-lncurses' " ` seem to be safer.

**Comment by wpoely86** [comment]
@hajgato Could you test the `libreadline` you need from #2206 ?

**Comment by boegel** [comment]
fixed in #2206
